Index: Makefile.inc
===================================================================
RCS file: /cvs/src/lib/libc/Makefile.inc,v
retrieving revision 1.14
diff -d -u -p -w Makefile.inc
--- Makefile.inc	3 Feb 2010 20:49:00 -0000	1.14
+++ Makefile.inc	26 Mar 2010 01:59:07 -0000
@@ -55,6 +55,7 @@ AINC+=  -nostdinc -idirafter ${DESTDIR}/usr/include
 .if (${YP:L} == "yes")
 .include "${LIBCSRCDIR}/yp/Makefile.inc"
 .endif
+.include "${LIBCSRCDIR}/mdns/Makefile.inc"
 
 CFLAGS+=-DNLS
 
Index: net/gethostnamadr.c
===================================================================
RCS file: /cvs/src/lib/libc/net/gethostnamadr.c,v
retrieving revision 1.73
diff -d -u -p -w net/gethostnamadr.c
--- net/gethostnamadr.c	18 Nov 2009 07:43:22 -0000	1.73
+++ net/gethostnamadr.c	26 Mar 2010 01:59:07 -0000
@@ -53,6 +53,7 @@
 #include <netinet/in.h>
 #include <arpa/inet.h>
 #include <arpa/nameser.h>
+#include <mdns.h>
 #include <netdb.h>
 #include <resolv.h>
 #include <stdio.h>
@@ -93,6 +94,7 @@ static int stayopen = 0;
 
 static void map_v4v6_address(const char *src, char *dst);
 static void map_v4v6_hostent(struct hostent *hp, char **bp, char *);
+static struct hostent *_mdns_gethtbyname(const char *);
 
 #ifdef RESOLVSORT
 static void addrsort(char **, int);
@@ -633,6 +635,15 @@ gethostbyname2(const char *name, int af)
 		case 'f':
 			hp = _gethtbyname2(name, af);
 			break;
+		case 'm':
+			/* MDNS only supports AF_INET and all hosts belong to
+			 * the .local domain */
+			if ((af != AF_INET)    ||
+			    (strlen(name) < 7) ||
+			    (strcmp(&name[strlen(name) - 6], ".local") != 0))
+				break;
+			
+			hp = _mdns_gethtbyname(name);
 		}
 	}
 	/* XXX h_errno not correct in all cases... */
@@ -906,6 +917,34 @@ _gethtbyaddr(const void *addr, socklen_t len, int af)
 			break;
 	_endhtent();
 	return (p);
+}
+
+static struct hostent *
+_mdns_gethtbyname(const char *name)
+{
+	static struct in_addr addr;
+	int r;
+	
+	r = mdns_lkup(name, &addr);
+	if (r == 0) {
+		h_errno = HOST_NOT_FOUND;
+		return NULL;
+	}
+	else if (r == -1) {
+		/* mdns error */
+		/* put some h_errno */
+		return NULL;
+	}
+	
+	/* host found */
+	h_addr_ptrs[0] = (char *) &addr;
+	host.h_name = (char *)name;
+	host.h_addr_list = h_addr_ptrs;
+	host.h_length = INADDRSZ;
+	host.h_addrtype = AF_INET;
+	host.h_aliases = NULL;
+	
+	return (&host);
 }
 
 #ifdef YP
Index: net/res_init.c
===================================================================
RCS file: /cvs/src/lib/libc/net/res_init.c,v
retrieving revision 1.40
diff -d -u -p -w net/res_init.c
--- net/res_init.c	5 Jun 2009 09:52:26 -0000	1.40
+++ net/res_init.c	26 Mar 2010 01:59:07 -0000
@@ -361,14 +361,17 @@ _res_init(int usercall)
 				    break;
 			    if ((*cp == '\0') || (*cp == '\n')) {
 				    if (sp) {
-					    if (*sp=='y' || *sp=='b' || *sp=='f')
+					    if (*sp=='y' || *sp=='b'
+						|| *sp=='f' || *sp =='m')
 						    _resp->lookups[n++] = *sp;
+
 					    sp = NULL;
 				    }
 				    break;
 			    } else if ((*cp == ' ') || (*cp == '\t') || (*cp == ',')) {
 				    if (sp) {
-					    if (*sp=='y' || *sp=='b' || *sp=='f')
+					    if (*sp=='y' || *sp=='b'
+						|| *sp=='f' || *sp=='m')
 						    _resp->lookups[n++] = *sp;
 					    sp = NULL;
 				    }
